File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        letter_counts = {}
    
        for letter in magazine:
            if letter in letter_counts:
                letter_counts[letter] += 1
            else:
                letter_counts[letter] = 1
        
        for letter in ransomNote:
            if letter in letter_counts and letter_counts[letter] > 0:
                letter_counts[letter] -= 1
            else:
                return False
        
        return True

File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map <Character, Integer> charFreq = new HashMap<>();
        
        for (char ch : magazine.toCharArray()){
            charFreq.put(ch, charFreq.getOrDefault(ch, 0) + 1);
        }

        for (char ch : ransomNote.toCharArray()){
            if (! charFreq.containsKey(ch)){
                return false;
            }
            charFreq.put(ch, charFreq.get(ch) - 1);
            if (charFreq.get(ch) < 0){
                return false;
            }
        }

        return true;
    }
}